src/main/java/com/impinjCtrl/ReportFormat.java

```diff
@@ -18,91 +18,119 @@
-              //  result.put("channelMhz", t.getChannelInMhz());
-            }
-            aggregateResult.add(result);
-
-            // log file
-            if(!TextUtils.isEmpty(mLogFileName)) {
-                mLogger.info(result.toJSONString());
+            String readingEpc = t.getEpc().toString().replace(" ", "").toLowerCase();
+            String epc = readingEpc;
+            Boolean isSlaveEpc = false;
+            Long ts = PropertyUtils.getTimestamp();
+            // Check if slave epc
+            if (ReaderController.mSlaveEpcMap != null && ReaderController.mSlaveEpcMap.get(readingEpc) != null) {
+                epc = ReaderController.mSlaveEpcMap.get(readingEpc).toString();
+                isSlaveEpc = true;
             }
-        }
-
-        if (!mIsDebugMode) {
-            // send tag data to api
-            JSONObject txData = new JSONObject();
-            txData.put("type", ReaderController.EVENT_TRANSFER_DATA);
-            txData.put("event", ReaderController.mEventId);
-            txData.put("payload", aggregateResult);
-
-            Request req = new Request.Builder()
-                    .url(PropertyUtils.getAPiHost() + "/api/socket/impinj?sid=" + ReaderController.mSocketId)
-                    .post(RequestBody.create(HttpClient.MEDIA_TYPE_JSON, txData.toJSONString()))
-                    .build();
-
-            mHttpClient.request(req, new Callback() {
-                public void onFailure(Call call, IOException e) {
-                    e.printStackTrace();
-                    System.out.println("update tag data fail");
+            if (mIsDebugMode) {
+                JSONObject result = new JSONObject();
+                try {
+                    result.put("epc", epc);
+                    result.put("timestamp", ts);
+                    result.put("isSlave", isSlaveEpc);
+                    result.put("antenna", t.getAntennaPortNumber());
+                    result.put("doppler", t.getRfDopplerFrequency());
+                    result.put("peakRssi", t.getPeakRssiInDbm());
+                    result.put("phase angel", t.getPhaseAngleInRadians());
+                }  catch (Exception e) {
+                    System.out.println("Assembling debug result error: " + e.getMessage());
                 }
-
-                public void onResponse(Call call, Response response) throws IOException {
-                    ResponseBody body = response.body();
-                    body.close();
+                System.out.println(result.toString());
+            } else {
+                JSONArray hashtableArray = new JSONArray();
+                JSONObject slaveStat = new JSONObject();
+                Integer lastValueIndex = -1;
+                // Check if entries exist
+                if (ReaderController.mRecordsHashTable.get(epc) != null) {
+                    hashtableArray = (JSONArray) ReaderController.mRecordsHashTable.get(epc);
```

Comment:
it's not a good idea to static values of the ReaderController, conventionally we will use getter / setter.
Although we will not get wrong value now, it's a extreme potential bug on this implementation when some day we want to new other controller and without destroy previous one well.

And other problem will cause: 
maybe we should design states now and refine the access flow, with these states we can send date without missing id or something else....